File: LunarML/exists.js
"use strict";
function _list(a) {
    let x = null;
    for (let i = a.length - 1; i >= 0; --i) {
        x = [a[i], x];
    }
    return x;
}
function _Match_tag() {}
_Match_tag.prototype.__isMLExn = true;
_Match_tag.prototype.name = "Match";
const _Match = new _Match_tag();
function _String_append(a, b) {
    if (a.length === 0) { return b; }
    if (b.length === 0) { return a; }
    const c = new Uint8Array(a.length + b.length);
    c.set(a);
    c.set(b, a.length);
    return c;
}
function _String_concat(xs) {
    let n = 0;
    let xs0 = xs;
    while (xs0 !== null) {
        const s = xs0[0];
        n += s.length;
        xs0 = xs0[1];
    }
    const a = new Uint8Array(n);
    let m = 0;
    while (xs !== null) {
        const s = xs[0];
        a.set(s, m);
        m += s.length;
        xs = xs[1];
    }
    return a;
}
function _wrap(f) {
    const F = function() {
        let result = f.apply(undefined, arguments); // [true, Result] | exists X. [false, MLFunction<Result, X>, X]
        while (!result[0]) {
            const g = result[1];
            if ("_MLTAIL_" in g) {
                result = g._MLTAIL_.apply(undefined, result[2]);
            } else {
                return g.apply(undefined, result[2]);
            }
        }
        return result[1];
    }
    F._MLTAIL_ = f;
    return F;
}
{
 const eq = function(a, b) {
  return a === b;
 };
 const revAppend = function(_1, ys) {
  let tmp22 = _1, tmp23 = ys;
  for (;;) {
   const _1$1 = tmp22, ys1 = tmp23;
   if (_1$1 === null) {
    return ys1;
   } else if (! (_1$1 === null)) {
    const tmp24 = _1$1[0];
    const tmp25 = _1$1[1];
    [tmp22, tmp23] = [tmp25, [tmp24, ys1]];
   } else {
    throw _Match;
   }
  }
 };
 const Io_tag = function(payload) {
  this.payload = payload;
 };
 Io_tag.prototype.__isMLExn = true;
 Io_tag.prototype.name = "Io";
 const BlockingNotSupported_tag = function() {
 };
 BlockingNotSupported_tag.prototype.__isMLExn = true;
 BlockingNotSupported_tag.prototype.name = "BlockingNotSupported";
 const BlockingNotSupported = new BlockingNotSupported_tag();
 const ClosedStream_tag = function() {
 };
 ClosedStream_tag.prototype.__isMLExn = true;
 ClosedStream_tag.prototype.name = "ClosedStream";
 const ClosedStream = new ClosedStream_tag();
 const NO_BUF = "NO_BUF";
 const outputAndFlush = function(_1, content) {
  const name = _1.writer.name;
  const writeVec = _1.writer.writeVec;
  const buffer = _1.buffer;
  const tmp22 = revAppend([content, buffer[0]], null);
  const content$PRIME = _String_concat(tmp22);
  if (writeVec.tag === "SOME") {
   const writeVec1 = writeVec.payload;
   writeVec1({base: content$PRIME, length: content$PRIME.length, start: 0});
   buffer[0] = null;
   return undefined;
  } else if (writeVec.tag === "NONE") {
   throw new Io_tag({cause: BlockingNotSupported, "function": Uint8Array.of(111, 117, 116, 112, 117, 116), name: name});
  } else {
   throw _Match;
  }
 };
 const closed = [false];
 const tmp = {tag: "SOME", payload: function(slice) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(119, 114, 105, 116, 101, 86, 101, 99), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return slice.length;
  }
 }};
 const tmp1 = {tag: "SOME", payload: function(slice) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(119, 114, 105, 116, 101, 65, 114, 114), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return slice.length;
  }
 }};
 const tmp2 = {tag: "SOME", payload: function(slice) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(119, 114, 105, 116, 101, 86, 101, 99, 78, 66), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return {tag: "SOME", payload: slice.length};
  }
 }};
 const tmp3 = {tag: "SOME", payload: function(slice) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(119, 114, 105, 116, 101, 65, 114, 114, 78, 66), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return {tag: "SOME", payload: slice.length};
  }
 }};
 const tmp4 = {tag: "SOME", payload: function(a) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(98, 108, 111, 99, 107), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return undefined;
  }
 }};
 const tmp5 = {tag: "SOME", payload: function(a) {
  if (closed[0]) {
   throw new Io_tag({cause: ClosedStream, "function": Uint8Array.of(99, 97, 110, 79, 117, 116, 112, 117, 116), name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62)});
  } else {
   return true;
  }
 }};
 const tmp6 = {tag: "NONE"};
 const tmp7 = {tag: "NONE"};
 const tmp8 = {tag: "NONE"};
 const tmp9 = {tag: "NONE"};
 const tmp10 = function(a) {
  closed[0] = true;
  return undefined;
 };
 const tmp11 = {block: tmp4, canOutput: tmp5, chunkSize: 1, close: tmp10, endPos: tmp7, getPos: tmp6, ioDesc: {tag: "NONE"}, name: Uint8Array.of(60, 110, 117, 108, 108, 87, 114, 62), setPos: tmp8, verifyPos: tmp9, writeArr: tmp1, writeArrNB: tmp3, writeVec: tmp, writeVecNB: tmp2};
 const tmp12 = [NO_BUF];
 const stdOut = [{buffer: [null], buffer_mode: tmp12, writer: tmp11}];
 let exists;
 exists = function(eq1) {
  return _wrap(function(a) {
   if (a[1] === null) {
    return [true, false];
   } else if (! (a[1] === null)) {
    const element = a[0];
    const tmp22 = a[1][0];
    const tmp23 = a[1][1];
    const tmp24 = eq1(element, tmp22);
    if (tmp24) {
     return [true, true];
    } else {
     const tmp25 = exists(eq1);
     return [false, tmp25, [[element, tmp23]]];
    }
   } else {
    throw _Match;
   }
  });
 };
 const tmp13 = exists(eq);
 const tmp14 = tmp13([2, _list([1, 2, 3])]);
 {
  let tmp22;
  if (tmp14) {
   tmp22 = Uint8Array.of(116, 114, 117, 101);
  } else if (! tmp14) {
   tmp22 = Uint8Array.of(102, 97, 108, 115, 101);
  } else {
   throw _Match;
  }
  const tmp23 = _String_append(tmp22, Uint8Array.of(10));
  outputAndFlush(stdOut[0], tmp23);
 }
 outputAndFlush(stdOut[0], Uint8Array.of(10));
 const tmp15 = exists(eq);
 const tmp16 = tmp15([3, _list([1, 2, 3])]);
 {
  let tmp22;
  if (tmp16) {
   tmp22 = Uint8Array.of(116, 114, 117, 101);
  } else if (! tmp16) {
   tmp22 = Uint8Array.of(102, 97, 108, 115, 101);
  } else {
   throw _Match;
  }
  const tmp23 = _String_append(tmp22, Uint8Array.of(10));
  outputAndFlush(stdOut[0], tmp23);
 }
 outputAndFlush(stdOut[0], Uint8Array.of(10));
 const tmp17 = exists(eq);
 const tmp18 = _list([1, 2]);
 const tmp19 = _list([3, 4]);
 let tmp20;
 {
  const tmp22 = revAppend(tmp18, null);
  tmp20 = revAppend(tmp22, tmp19);
 }
 const tmp21 = tmp17([2, tmp20]);
 {
  let tmp22;
  if (tmp21) {
   tmp22 = Uint8Array.of(116, 114, 117, 101);
  } else if (! tmp21) {
   tmp22 = Uint8Array.of(102, 97, 108, 115, 101);
  } else {
   throw _Match;
  }
  const tmp23 = _String_append(tmp22, Uint8Array.of(10));
  outputAndFlush(stdOut[0], tmp23);
 }
 outputAndFlush(stdOut[0], Uint8Array.of(10));
}

File: LunarML/exists.lua
local error = error
local getmetatable = getmetatable
local setmetatable = setmetatable
local string = string
local string_format = string.format
local table = table
local table_concat = table.concat
local _exn_meta = {}
function _exn_meta:__tostring()
  local traceback = self.traceback
  if traceback then
    traceback = "\n" .. traceback
  else
    traceback = ""
  end
  return string_format("%s: %s%s", self.location or "<no location info>", self.tag[1], traceback)
end
local _Match_tag = { "Match" }
local _Match = setmetatable({ tag = _Match_tag }, _exn_meta)
local _Fail_tag = { "Fail" }
local function _raise(x, location)
  local e
  if getmetatable(x) == _exn_meta and location ~= nil then
    local traceback = debug.traceback(nil, 2)
    e = setmetatable({ tag = x.tag, payload = x.payload, location = location, traceback = traceback }, _exn_meta)
  else
    e = x
  end
  error(e, 1)
end
local function _VectorOrArray_fromList(xs)
  local t = {}
  local n = 0
  while xs ~= nil do
    n = n + 1
    t[n] = xs[1]
    xs = xs[2]
  end
  t.n = n
  return t
end
local revAppend, outputAndFlush, stdOut, tmp
do
  local eq = function(a, b)
    return a == b
  end
  revAppend = function(_1, ys)
    local tmp1, tmp2 = _1, ys
    ::cont::
    local _1_1, ys1 = tmp1, tmp2
    if _1_1 == nil then
      return ys1
    elseif _1_1 ~= nil then
      local tmp3 = _1_1[1]
      local tmp4 = _1_1[2]
      tmp1 = tmp4
      tmp2 = {tmp3, ys1}
      goto cont
    else
      _raise(_Match, "list.sml:60:5")
    end
  end
  local stdout = _ENV.io.stdout
  local Io_tag = {"Io"}
  local BlockingNotSupported = setmetatable({tag = {"BlockingNotSupported"}}, _exn_meta)
  local LINE_BUF = "LINE_BUF"
  outputAndFlush = function(_1, content)
    if _1.tag == "LUA_WRITABLE" then
      local writable = _1.payload.writable
      local name = _1.payload.name
      local tmp1, tmp2 = writable:write(content)
      if not tmp1 then
        _raise(setmetatable({tag = Io_tag, payload = {cause = setmetatable({tag = _Fail_tag, payload = tmp2}, _exn_meta), ["function"] = "output", name = name}}, _exn_meta), "text-io.sml:290:43")
      else
        writable:flush()
        return nil
      end
    elseif _1.tag == "PRIM_WRITER" then
      local name = _1.payload.writer.name
      local writeVec = _1.payload.writer.writeVec
      local buffer = _1.payload.buffer
      local tmp1 = revAppend({content, buffer[1]}, nil)
      local tmp2 = _VectorOrArray_fromList(tmp1)
      local content_PRIME = table_concat(tmp2)
      if writeVec.tag == "SOME" then
        local writeVec1 = writeVec.payload
        writeVec1({base = content_PRIME, length = #content_PRIME, start = 0})
        buffer[1] = nil
        return nil
      elseif writeVec.tag == "NONE" then
        _raise(setmetatable({tag = Io_tag, payload = {cause = BlockingNotSupported, ["function"] = "output", name = name}}, _exn_meta), "text-io.sml:397:26")
      else
        _raise(_Match, "text-io.sml:395:14")
      end
    else
      _raise(_Match, "text-io.sml:392:9")
    end
  end
  stdOut = {tag = "LUA_WRITABLE", payload = {buffer_mode = {LINE_BUF}, name = "<stdout>", writable = stdout}}
  local function exists(eq1)
    return function(a)
      if a[2] == nil then
        return false
      elseif a[2] ~= nil then
        local element = a[1]
        local tmp1 = a[2][1]
        local tmp2 = a[2][2]
        local tmp3 = eq1(element, tmp1)
        if tmp3 then
          return true
        end
        local tmp4 = exists(eq1)
        return tmp4({element, tmp2})
      else
        _raise(_Match, "exists.sml:3:5")
      end
    end
  end
  local tmp1 = exists(eq)
  local tmp2 = tmp1({2, {1, {2, {3, nil}}}})
  do
    local tmp3
    if tmp2 then
      tmp3 = "true"
    elseif not tmp2 then
      tmp3 = "false"
    else
      _raise(_Match, "exists.sml:11:5")
    end
    outputAndFlush(stdOut, tmp3 .. "\n")
  end
  outputAndFlush(stdOut, "\n")
  local tmp3 = exists(eq)
  local tmp4 = tmp3({3, {1, {2, {3, nil}}}})
  do
    local tmp5
    if tmp4 then
      tmp5 = "true"
    elseif not tmp4 then
      tmp5 = "false"
    else
      _raise(_Match, "exists.sml:11:5")
    end
    outputAndFlush(stdOut, tmp5 .. "\n")
  end
  outputAndFlush(stdOut, "\n")
  tmp = exists(eq)
end
local tmp1 = {1, {2, nil}}
local tmp2 = {3, {4, nil}}
local tmp3
do
  local tmp4 = revAppend(tmp1, nil)
  tmp3 = revAppend(tmp4, tmp2)
end
local tmp4 = tmp({2, tmp3})
do
  local tmp5
  if tmp4 then
    tmp5 = "true"
  elseif not tmp4 then
    tmp5 = "false"
  else
    _raise(_Match, "exists.sml:11:5")
  end
  outputAndFlush(stdOut, tmp5 .. "\n")
end
outputAndFlush(stdOut, "\n")
